crates/volta-core/src/tool/yarn/mod.rs

```diff
@@ -0,0 +1,77 @@
+use std::fmt::{self, Display};
+
+use super::{debug_already_fetched, info_fetched, info_installed, info_pinned, Tool};
+use crate::error::ErrorDetails;
+use crate::session::Session;
+use crate::style::tool_version;
+use semver::Version;
+use volta_fail::Fallible;
+
+mod fetch;
+mod resolve;
+mod serial;
+
+pub use resolve::resolve;
+
+/// The Tool implementation for fetching and installing Yarn
+#[derive(Debug)]
+pub struct Yarn {
+    pub(super) version: Version,
+}
+
+impl Yarn {
+    pub fn new(version: Version) -> Self {
+        Yarn { version }
+    }
+
+    pub(crate) fn fetch_internal(&self, session: &mut Session) -> Fallible<()> {
+        let inventory = session.inventory()?;
+        if inventory.yarn.contains(&self.version) {
+            debug_already_fetched(self);
+            return Ok(());
+        }
+
+        fetch::fetch(&self.version, session.hooks()?.yarn())?;
+        session
+            .inventory_mut()?
+            .yarn
+            .versions
+            .insert(self.version.clone());
+
+        Ok(())
+    }
+}
+
+impl Tool for Yarn {
+    fn fetch(self, session: &mut Session) -> Fallible<()> {
```

Comment:
I note that the design of the `Tool` trait is what necessitates passing down `Session` instead, so we'd need to push that refactor back up the chain. However, I think it's likely worth it for all the reasons mentioned above. If the `Tool` trait can become entirely ignorant of `Session`, that'd be a huge win in my book.

Reply:
I think it's a bit of a balance, given that the different implementations use different subsets of the available data. If we restructure it to pass in exactly what is needed, then the parameters on the Trait method need to be the superset that contains the necessary data for _any_ of the implementations. It also means if we, in the future, want to add some additional functionality that requires other data off of the `Session`, then we have to plumb that through the trait and each of the implementations.

For example, one follow-on refactor that we should be able to do is to generate the `PlatformSpec` and call `npm view` directly in the `package::resolve` module, instead of the current design which requires calling out to the `npm` shim. That will require us to pass along the result of `current_platform` to the `resolve` method, even though most of the resolvers don't actually need to know about that.

While I agree the `Session` is a "bag of everything", I think this refactor has removed a lot of the behavior, which lets it be more of a "state of the world" object, which I think is a bit more reasonable to pass around.

Reply:
Yeah, that's also a fair take. In situations where I find that to be the case, it makes me wonder if I've actually over-abstracted. (In general, I'm starting to lean toward the conclusion that it's *very* difficult to properly abstract over *all* of runtimes, package managers, and packages.)

I'm happy to defer to you on this, and if/when we have pain in the future we can make a refactor-to-make-the-hard-thing-easy at that point.

Reply:
Yeah, it's definitely possible that we're still over-abstracted even given how much has been pulled out, but I think waiting to see if extra pain points come up is a good approach. We definitely need to make sure we call out as soon as something doesn't feel right, and not just shoe-horn it into working, so that we can refactor quicker and smaller, and not let it grow until it's a tangled mess.

Reply:
Yep, I think we're agreed on approach here! 👍